File: backend/rollingpebble/api/storage.py
from __future__ import annotations

from fastapi import APIRouter

from rollingpebble.api.context import AppServices, open_folder, service_call
from rollingpebble.models import (
    StorageCleanupPlanResponse,
    StorageCleanupPreviewRequest,
    StorageCleanupRunRequest,
    StorageCleanupRunResponse,
    StorageMigrateRootRequest,
    StorageMigrateRootResponse,
    StorageOpenModelRequest,
    StorageOpenOtherRequest,
    StorageOpenRuntimeRequest,
    StorageUsageResponse,
)


def create_storage_router(services: AppServices) -> APIRouter:
    router = APIRouter()

    @router.post("/api/storage/projects/open-folder")
    def open_projects_folder() -> dict[str, str]:
        folder = services.storage.layout.projects_root
        folder.mkdir(parents=True, exist_ok=True)
        return open_folder(folder)

    @router.post("/api/storage/open-folder")
    def open_storage_folder() -> dict[str, str]:
        folder = services.storage.layout.app_root
        folder.mkdir(parents=True, exist_ok=True)
        return open_folder(folder)

    @router.get("/api/storage/usage", response_model=StorageUsageResponse)
    def storage_usage() -> StorageUsageResponse:
        return services.storage.usage()

    @router.post("/api/storage/migrate-root", response_model=StorageMigrateRootResponse)
    def storage_migrate_root(request: StorageMigrateRootRequest) -> StorageMigrateRootResponse:
        def migrate() -> StorageMigrateRootResponse:
            result = services.storage.migrate_root(request.root_id, request.target_path)
            services.apply_storage_layout()
            return result

        return service_call(migrate)

    @router.post("/api/storage/models/open-folder")
    def open_model_folder(request: StorageOpenModelRequest) -> dict[str, str]:
        return service_call(lambda: open_folder(services.storage.model_item_path(request.model_id)))

    @router.post("/api/storage/runtimes/open-folder")
    def open_runtime_folder(request: StorageOpenRuntimeRequest) -> dict[str, str]:
        return service_call(lambda: open_folder(services.storage.runtime_item_path(request.runtime_id)))

    @router.post("/api/storage/other/open-folder")
    def open_other_folder(request: StorageOpenOtherRequest) -> dict[str, str]:
        return service_call(lambda: open_folder(services.storage.other_item_open_path(request.relative_path)))

    @router.post("/api/storage/cleanup/preview", response_model=StorageCleanupPlanResponse)
    def storage_cleanup_preview(request: StorageCleanupPreviewRequest) -> StorageCleanupPlanResponse:
        return service_call(lambda: services.storage.preview(request))

    @router.post("/api/storage/cleanup/run", response_model=StorageCleanupRunResponse)
    def storage_cleanup_run(request: StorageCleanupRunRequest) -> StorageCleanupRunResponse:
        return service_call(lambda: services.storage.run(request), not_found=(FileNotFoundError, KeyError))

    return router

File: frontend/src/shared/api/storage.ts
import { request } from "./request.js";
import type {
    StorageCleanupPlan,
    StorageCleanupRunResult,
    StorageCleanupTarget,
    StorageMigrateRootResult,
    StorageUsage,
} from "./types.js";

export const storageUsage = (): Promise<StorageUsage> => request<StorageUsage>("/api/storage/usage");

export const storageCleanupPreview = (
    payload: {
        targets: StorageCleanupTarget[];
        older_than_days?: number | null;
        project_ids?: string[];
        model_ids?: string[];
        runtime_ids?: string[];
        other_paths?: string[];
    },
): Promise<StorageCleanupPlan> =>
    request<StorageCleanupPlan>("/api/storage/cleanup/preview", { method: "POST", body: JSON.stringify(payload) });

export const openStorageFolder = (): Promise<{ status: string; path: string }> =>
    request<{ status: string; path: string }>("/api/storage/open-folder", { method: "POST" });

export const migrateStorageRoot = (
    payload: { root_id: "projects" | "models" | "cache"; target_path: string },
): Promise<StorageMigrateRootResult> =>
    request<StorageMigrateRootResult>("/api/storage/migrate-root", { method: "POST", body: JSON.stringify(payload) });

export const openModelFolder = (modelId: string): Promise<{ status: string; path: string }> =>
    request<{ status: string; path: string }>("/api/storage/models/open-folder", {
        method: "POST",
        body: JSON.stringify({ model_id: modelId }),
    });

export const openRuntimeFolder = (runtimeId: string): Promise<{ status: string; path: string }> =>
    request<{ status: string; path: string }>("/api/storage/runtimes/open-folder", {
        method: "POST",
        body: JSON.stringify({ runtime_id: runtimeId }),
    });

export const openOtherFolder = (relativePath: string): Promise<{ status: string; path: string }> =>
    request<{ status: string; path: string }>("/api/storage/other/open-folder", {
        method: "POST",
        body: JSON.stringify({ relative_path: relativePath }),
    });

export const storageCleanupRun = (
    payload: { plan_id: string; entry_ids?: string[] | null },
): Promise<StorageCleanupRunResult> =>
    request<StorageCleanupRunResult>("/api/storage/cleanup/run", { method: "POST", body: JSON.stringify(payload) });
